packages/flutter/lib/src/widgets/editable_text.dart

```diff
@@ -2236,6 +2235,20 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
       }
     }
 
+    if (_value.selection != value.selection) {
```

Comment:
I believe `RenderEditable` also calls `_handleSelectionChanged` when handling raw keyboard events (i.e. in `RenderEditable._handleMovement`), after calling `_formatAndSave`. Sounds like `widget.onSelectionChanged` is going to be called twice in this scenario?

Reply:
It looks like we should call handleSelectionChange first and then sets textEditingValue in renderEditable, I am quite curious though, why do we need to set textEditingValue in the first place?
For _handleShortcuts, LogicalKeyboardKey.keyA only calls _handleselectionchange but not setting the textEditingValue

Reply:
Traced it to https://github.com/flutter/flutter/pull/44130 where the `textSelectionDelegate.textEditingValue = textSelectionDelegate.textEditingValue.copyWith(selection: newSelection);` was first added. It looks like we want to send the selection change to the engine.

Reply:
I didnt see any code path that will send the change to the enigne. Do you know where does that happen?

Reply:
The `_value = value` line does it (in one of the listeners).

https://github.com/flutter/flutter/blob/8cd03a2e3649d9fc92442f2ea04417fcf30ebb70/packages/flutter/lib/src/widgets/editable_text.dart#L2225-L2237

Reply:
ah i thought _value was just a variable, that make sense, but should we also apply this to handle shortcut as well?

Reply:
Maybe we should change it to `widget.controller.value = value` so this line doesn't get overlooked. 

You mean this part?
https://github.com/flutter/flutter/blob/be1325f18b5e557641fa48e0e40fa694fae131e1/packages/flutter/lib/src/rendering/editable.dart#L805-L814

Yeah looks like it needs to call the `textEditingValue` setter here too. It feels the callbacks in `RenderEditable` are getting a bit out of hand.  I think we're moving the keyboard events handling out of the `RenderEditable` class? /cc @justinmc

Reply:
That's the plan.  This PR is reminding me how important it is.  These sorts of changes shouldn't be so confusing.